### ISSUE TITLE
 java.lang.RuntimeException: whoops, better handle OTHER null

### ISSUE THREAD

**Issue by guismart** [uber/NullAway#366]
Encounter error when use Gradle to compile google android code with Error Prone plugin:
```
src/com/android/providers/settings/SettingsProvider.java:1487: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
return settingsState.new Setting(ssaidSetting) {
^

error-prone version: 2.3.3
BugPattern: NullAway

 Stack Trace:
 java.lang.RuntimeException: whoops, better handle OTHER null
at com.uber.nullaway.NullAway.mayBeNullExpr(NullAway.java:1952)
at com.uber.nullaway.NullAway.handleInvocation(NullAway.java:1352)
at com.uber.nullaway.NullAway.matchNewClass(NullAway.java:354)
at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:433)
at com.google.errorprone.scanner.ErrorProneScanner.visitNewClass(ErrorProneScanner.java:759)
at com.google.errorprone.scanner.ErrorProneScanner.visitNewClass(ErrorProneScanner.java:150)
at com.sun.tools.javac.tree.JCTree$JCNewClass.accept(JCTree.java:1705)
```

Gradle compilation passed without error-prone plugin, but failed if plugin integrated.

Relevant code fragment:

```
private Setting mascaradeSsaidSetting(SettingsState settingsState, Setting ssaidSetting) {
    if (ssaidSetting != null) {
        return settingsState.new Setting(ssaidSetting) {
              @override
               public int getKey() {
                    final int userId = getUserIdFromKey(super.getKey());
                    return makeKey(SETTINGS_TYPE_SECURE, userId);
                  }

             @Override
             public String getName() {
                 return Settings.Secure.ANDROID_ID;
              }
          };
      }
    return null;
}
```

**Comment by msridhar** [comment]
@shas19 @lazaroclapp either of you have time to look at this one?

**Comment by shas19** [comment]
Yeah, I will check it.

**Comment by shas19** [comment]
Apparently, there are two types of issues I can see.  

Test 1:
```
  class AnotherClass{
    class Inner{
      int x;
    }
  }

  AnotherClass.Inner someMethod(){
    AnotherClass anotherClass = new AnotherClass();
    return anotherClass.new Inner(){
      int getX(){ return x;}
    };
  }
```

Java compiler parses the invocation of `Inner` constructor with parameter `<*nullchk*>anotherClass`.
Since `AccessPathNullnessPropagation#visitNullChk` is not expected to occur, but it is occuring, that is causing one type of crash. 

Test 2:
```
  class AnotherClass{
    class Inner{
      int x;
      Inner(int __x) {x = __x;}
      int getX(){return 3;}
    }
  }

  AnotherClass.Inner someMethod(){
    AnotherClass anotherClass = new AnotherClass();
    return anotherClass.new Inner(4){
      @Override
      int getX(){ return 5;}
    };
  }
```

This test is similar to the test in the issue. Now again, Java compiler parses the invocation as having parameters `<*nullchk*>anotherClass`, 4. In this case while unboxing check, `Nullaway#mayBeNullExpr` is called on the `<*nullchk*>anotherClass`. Since we do not expect `tree.getKind` of the tree to be `OTHER`, this error is thrown. Here the class of the tree is `class com.sun.tools.javac.tree.JCTree$JCUnary`.  Basically `<*nullchk*>` is the unary operation on `anotherClass`. 

@msridhar Why did we assume that `AccessPathNullnessPropagation#visitNullChk` is never called?

**Comment by msridhar** [comment]
@shas19 thanks for tracking this down!  I don't know why we made that assumption.  I just had never seen that type of node before.  I'm not really sure what the node does, but my best guess is its nullability should be the same as the nullability of its child expression.  In other words, it should be treated exactly like a type cast.  So concretely, here is what we should do:

1. In `AccessPathNullnessPropagation`, have `visitNullChk` be exactly like `visitTypeCast`:

https://github.com/uber/NullAway/blob/f1307b06a1324b8b65d0fae269ff6c74d54b194d/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java#L801-L807

2. In `NullAway` we have a routine `stripParensAndCasts` called at the beginning of `mayBeNullExpr`.  This routine should also strip away `NullChk` operators.

Does that make sense?

**Comment by msridhar** [comment]
@guismart FYI the fix for this issue is in release 0.7.9.  Thanks for reporting!
